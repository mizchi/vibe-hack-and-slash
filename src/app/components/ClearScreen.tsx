import React, { useEffect } from "react";
import { Box, useStdout } from "ink";

type Props = {
  children: React.ReactNode;
};

// 画面をクリアするラッパーコンポーネント
export const ClearScreen: React.FC<Props> = ({ children }) => {
  const { stdout } = useStdout();

  useEffect(() => {
    // コンポーネントマウント時に画面をクリア
    if (stdout) {
      // カーソルを一番上に移動
      stdout.write('\x1b[H');
      // 画面をクリア
      stdout.write('\x1b[2J');
      // カーソルを一番上に移動（再度）
      stdout.write('\x1b[H');
    }
  }, [stdout]);

  return (
    <Box flexDirection="column" width="100%" height={40}>
      {children}
    </Box>
  );
};