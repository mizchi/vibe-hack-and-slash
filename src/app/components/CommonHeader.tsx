import React from "react";
import { Box, Text } from "ink";
import type { Session } from "../../core/types.ts";
import { formatGold } from "../../core/item-value.ts";

type Props = {
  session: Session;
  currentView: "battle" | "equipment";
  gameSpeed?: 0 | 1 | 3 | 5;
  battleStatus?: {
    isInBattle: boolean;
    currentMonster?: {
      name: string;
      level: number;
      healthPercent: number;
    };
  };
};

export const CommonHeader: React.FC<Props> = ({ session, currentView, gameSpeed = 1, battleStatus }) => {
  return (
    <Box borderStyle="double" padding={1} marginBottom={1}>
      <Box flexDirection="column" width="100%">
        {/* タイトル行 */}
        <Box justifyContent="space-between">
          <Box>
            <Text bold color="cyan">Hack & Slash</Text>
            <Text> - </Text>
            <Text>{session.player.class} Lv.{session.player.level}</Text>
          </Box>
          <Box>
            <Text color="yellow">[Gold: {formatGold(session.player.gold)}]</Text>
            {gameSpeed !== undefined && (
              <>
                <Text> </Text>
                <Text color={gameSpeed === 0 ? "red" : gameSpeed > 1 ? "green" : "white"}>
                  [Speed: x{gameSpeed}]
                </Text>
              </>
            )}
          </Box>
        </Box>
        
        {/* ナビゲーションと戦闘状態 */}
        <Box marginTop={1} justifyContent="space-between">
          <Box>
            <Text dimColor>Tab: </Text>
            <Text color={currentView === "battle" ? "cyan" : "gray"} bold={currentView === "battle"}>
              [戦闘詳細]
            </Text>
            <Text dimColor> | </Text>
            <Text color={currentView === "equipment" ? "cyan" : "gray"} bold={currentView === "equipment"}>
              [装備管理]
            </Text>
          </Box>
          
          {/* 戦闘状態表示 */}
          {battleStatus && (
            <Box>
              {battleStatus.isInBattle ? (
                <>
                  <Text color="red">[戦闘中] </Text>
                  {battleStatus.currentMonster && (
                    <Text>
                      {battleStatus.currentMonster.name} Lv.{battleStatus.currentMonster.level}
                      <Text dimColor> (HP: {battleStatus.currentMonster.healthPercent}%)</Text>
                    </Text>
                  )}
                </>
              ) : (
                <Text color="green">[拠点 - 安全]</Text>
              )}
            </Box>
          )}
        </Box>
      </Box>
    </Box>
  );
};