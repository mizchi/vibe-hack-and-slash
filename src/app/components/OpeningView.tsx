import React, { useState } from "react";
import { Box, Text, useInput } from "ink";
import type { PlayerClass } from "../../core/types.ts";

type Props = {
  onClassSelected: (playerClass: PlayerClass) => void;
};

export const OpeningView: React.FC<Props> = ({ onClassSelected }) => {
  const classes: PlayerClass[] = ["Warrior", "Mage", "Rogue", "Paladin"];
  const [selectedIndex, setSelectedIndex] = useState(0);
  
  const classDescriptions = {
    Warrior: "高い体力と防御力を持つ戦士",
    Mage: "強力な魔法と高いマナを持つ魔法使い",
    Rogue: "高いクリティカル率と攻撃速度を持つ盗賊",
    Paladin: "攻撃と防御のバランスが取れた聖騎士"
  };

  const classStats = {
    Warrior: { hp: "120", atk: "15", def: "5", mana: "30" },
    Mage: { hp: "80", atk: "8", def: "0", mana: "100" },
    Rogue: { hp: "90", atk: "12", def: "2", mana: "50" },
    Paladin: { hp: "110", atk: "12", def: "8", mana: "60" }
  };

  useInput((input, key) => {
    if (key.upArrow) {
      setSelectedIndex(Math.max(0, selectedIndex - 1));
    }
    if (key.downArrow) {
      setSelectedIndex(Math.min(classes.length - 1, selectedIndex + 1));
    }
    if (key.return) {
      onClassSelected(classes[selectedIndex]);
    }
  });

  return (
    <Box flexDirection="column">
      <Box borderStyle="double" padding={1} marginBottom={1}>
        <Text bold>⚔️  Hack & Slash - クラス選択</Text>
      </Box>
      
      <Box flexDirection="row">
        <Box flexDirection="column" width="60%">
          {classes.map((cls, index) => (
            <Box key={cls} marginBottom={1}>
              <Text color={index === selectedIndex ? "cyan" : undefined}>
                {index === selectedIndex ? "▶ " : "  "}
                {cls}: {classDescriptions[cls]}
              </Text>
            </Box>
          ))}
        </Box>
        
        <Box flexDirection="column" width="40%" borderStyle="single" padding={1}>
          <Text bold underline>
            {classes[selectedIndex]} のステータス
          </Text>
          <Text>HP: {classStats[classes[selectedIndex]].hp}</Text>
          <Text>攻撃力: {classStats[classes[selectedIndex]].atk}</Text>
          <Text>防御力: {classStats[classes[selectedIndex]].def}</Text>
          <Text>マナ: {classStats[classes[selectedIndex]].mana}</Text>
        </Box>
      </Box>
      
      <Box marginTop={1}>
        <Text dimColor>↑↓: 選択 | Enter: 決定</Text>
      </Box>
    </Box>
  );
};