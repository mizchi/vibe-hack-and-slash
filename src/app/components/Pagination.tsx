import React from "react";
import { Box, Text } from "ink";

type Props = {
  currentPage: number;
  totalPages: number;
  itemsPerPage: number;
  totalItems: number;
  startIndex: number;
  endIndex: number;
};

export const Pagination: React.FC<Props> = ({
  currentPage,
  totalPages,
  itemsPerPage,
  totalItems,
  startIndex,
  endIndex,
}) => {
  if (totalPages <= 1) return null;

  // ページインジケーターの生成
  const generatePageIndicators = () => {
    const indicators = [];
    const maxVisiblePages = 5;
    
    let start = Math.max(0, currentPage - Math.floor(maxVisiblePages / 2));
    let end = Math.min(totalPages, start + maxVisiblePages);
    
    // 開始位置の調整
    if (end - start < maxVisiblePages) {
      start = Math.max(0, end - maxVisiblePages);
    }
    
    // 最初のページ
    if (start > 0) {
      indicators.push(
        <Text key="first" dimColor>1</Text>,
        <Text key="dots-start" dimColor>...</Text>
      );
    }
    
    // ページ番号
    for (let i = start; i < end; i++) {
      indicators.push(
        <Text
          key={i}
          color={i === currentPage ? "cyan" : undefined}
          bold={i === currentPage}
          dimColor={i !== currentPage}
        >
          {i + 1}
        </Text>
      );
    }
    
    // 最後のページ
    if (end < totalPages) {
      indicators.push(
        <Text key="dots-end" dimColor>...</Text>,
        <Text key="last" dimColor>{totalPages}</Text>
      );
    }
    
    return indicators;
  };

  return (
    <Box>
      <Text dimColor>[</Text>
      {generatePageIndicators().map((indicator, index) => (
        <React.Fragment key={`indicator-${index}`}>
          {index > 0 && <Text dimColor> </Text>}
          {indicator}
        </React.Fragment>
      ))}
      <Text dimColor>]</Text>
      <Text dimColor> ({startIndex + 1}-{Math.min(endIndex, totalItems)}/{totalItems})</Text>
    </Box>
  );
};