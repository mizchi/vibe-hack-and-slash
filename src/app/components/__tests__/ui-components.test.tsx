import React from "react";
import { render } from "ink-testing-library";
import { Text, Box } from "ink";
import { describe, it, expect, beforeAll } from "vitest";

// コンポーネントのインポート
import { ClearScreen } from "../ClearScreen";
import { CommonHeader } from "../CommonHeader";
import { OpeningView } from "../OpeningView";
import { Pagination } from "../Pagination";

// モックデータ
import type { Session, PlayerClass } from "../../../core/types";

// テスト環境設定
beforeAll(() => {
  process.stdin.isTTY = true;
  process.stdout.isTTY = true;
});

// モックセッションデータ
const createMockSession = (): Session => ({
  id: "test-session" as any,
  player: {
    id: "test-player" as any,
    class: "Warrior" as PlayerClass,
    level: 5 as any,
    currentHealth: 100 as any,
    currentMana: 50 as any,
    experience: 0 as any,
    gold: 100 as any,
    equipment: new Map(),
    baseAttributes: {
      strength: 10 as any,
      intelligence: 5 as any,
      dexterity: 8 as any,
      vitality: 10 as any
    },
    name: "テストプレイヤー",
    baseStats: {
      maxHealth: 100 as any,
      baseDamage: 10 as any,
      criticalChance: 0.1,
      criticalDamage: 1.5,
      lifeSteal: 0,
      mpRegen: 2 as any,
      elementModifier: {
        Physical: 1,
        Arcane: 0,
        Fire: 0,
        Lightning: 0,
        Holy: 0
      }
    },
    inventory: [],
    skills: [],
    skillCooldowns: new Map(),
    skillTimers: new Map(),
    activeBuffs: [],
    elementResistance: {
      Physical: 0,
      Arcane: 0,
      Fire: 0,
      Lightning: 0,
      Holy: 0
    }
  },
  defeatedCount: 0,
  wave: 3,
  state: "InProgress" as const,
  startedAt: new Date()
});

describe("UI Components", () => {
  describe("ClearScreen", () => {
    it("画面をクリアするエスケープシーケンスを出力", () => {
      const { lastFrame } = render(<ClearScreen>
        <Text>Child Content</Text>
      </ClearScreen>);
      // ClearScreenは特殊な文字を出力するため、存在確認のみ
      expect(lastFrame).toBeDefined();
    });
  });

  describe("CommonHeader", () => {
    it("基本的なヘッダー表示", () => {
      const mockSession = createMockSession();
      const { lastFrame } = render(
        <CommonHeader
          session={mockSession}
          currentView="battle"
          gameSpeed={1}
        />
      );
      
      const output = lastFrame();
      expect(output).toContain("Hack & Slash");
      expect(output).toContain("Warrior Lv.5");
      expect(output).toContain("[Gold: 100]");
      expect(output).toContain("[Speed: x1]");
    });

    it("スナップショット", () => {
      const mockSession = createMockSession();
      const { lastFrame } = render(
        <CommonHeader
          session={mockSession}
          currentView="equipment"
          gameSpeed={3}
        />
      );
      
      expect(lastFrame()).toMatchSnapshot();
    });
  });

  describe("OpeningView", () => {
    it("クラス選択画面の表示", () => {
      const onClassSelected = () => {};
      const { lastFrame } = render(
        <OpeningView onClassSelected={onClassSelected} />
      );
      
      const output = lastFrame();
      expect(output).toContain("Hack & Slash");
      expect(output).toContain("Warrior");
      expect(output).toContain("Mage");
      expect(output).toContain("Rogue");
      expect(output).toContain("Paladin");
    });

    it("スナップショット", () => {
      const onClassSelected = () => {};
      const { lastFrame } = render(
        <OpeningView onClassSelected={onClassSelected} />
      );
      
      expect(lastFrame()).toMatchSnapshot();
    });
  });

  describe("Pagination", () => {
    it("ページネーションの表示", () => {
      const { lastFrame } = render(
        <Pagination
          currentPage={0}
          totalPages={5}
          itemsPerPage={10}
          totalItems={45}
          startIndex={0}
          endIndex={10}
        />
      );
      
      const output = lastFrame();
      expect(output).toContain("[");
      expect(output).toContain("]");
      expect(output).toContain("1-10/45");
    });

    it("複数ページの表示", () => {
      const { lastFrame } = render(
        <Pagination
          currentPage={2}
          totalPages={5}
          itemsPerPage={10}
          totalItems={50}
          startIndex={20}
          endIndex={30}
        />
      );
      
      const output = lastFrame();
      expect(output).toContain("3"); // 現在のページ（0-indexed なので 2 = 3ページ目）
      expect(output).toContain("21-30/50");
    });

    it("単一ページの場合は表示しない", () => {
      const { lastFrame } = render(
        <Pagination
          currentPage={0}
          totalPages={1}
          itemsPerPage={10}
          totalItems={5}
          startIndex={0}
          endIndex={5}
        />
      );
      
      expect(lastFrame()).toBe("");
    });

    it("スナップショット", () => {
      const { lastFrame } = render(
        <Pagination
          currentPage={2}
          totalPages={10}
          itemsPerPage={10}
          totalItems={100}
          startIndex={20}
          endIndex={30}
        />
      );
      
      expect(lastFrame()).toMatchSnapshot();
    });
  });
});

// モックデータを使った複雑なコンポーネントのテスト
describe("Complex UI Components", () => {
  it("カスタムコンポーネントの組み合わせ", () => {
    const StatusBar: React.FC<{ hp: number; mp: number }> = ({ hp, mp }) => (
      <Box flexDirection="row" gap={2}>
        <Text color="red">HP: {hp}</Text>
        <Text color="blue">MP: {mp}</Text>
      </Box>
    );

    const GameMenu: React.FC<{ items: string[]; selected: number }> = ({ items, selected }) => (
      <Box flexDirection="column" borderStyle="single" padding={1}>
        {items.map((item, i) => (
          <Text key={i} color={i === selected ? "cyan" : undefined}>
            {i === selected ? "> " : "  "}{item}
          </Text>
        ))}
      </Box>
    );

    const ComplexUI: React.FC = () => (
      <Box flexDirection="column" gap={1}>
        <StatusBar hp={100} mp={50} />
        <GameMenu items={["Attack", "Defend", "Item", "Run"]} selected={0} />
      </Box>
    );

    const { lastFrame } = render(<ComplexUI />);
    const output = lastFrame();
    
    expect(output).toContain("HP: 100");
    expect(output).toContain("MP: 50");
    expect(output).toContain("> Attack");
    expect(output).toContain("  Defend");
    expect(output).toContain("  Item");
    expect(output).toContain("  Run");
  });
});

// ヘルパー関数のテスト
describe("UI Helpers", () => {
  it("ヘルスバーの表示", () => {
    const HealthBar: React.FC<{ current: number; max: number; width?: number }> = ({
      current,
      max,
      width = 10
    }) => {
      const percentage = Math.round((current / max) * 100);
      const filled = Math.round((current / max) * width);
      const empty = width - filled;

      return (
        <Box>
          <Text>HP: </Text>
          <Text color="red">{"█".repeat(filled)}</Text>
          <Text dimColor>{"░".repeat(empty)}</Text>
          <Text> {current}/{max} ({percentage}%)</Text>
        </Box>
      );
    };

    const { lastFrame } = render(<HealthBar current={75} max={100} />);
    const output = lastFrame();
    
    expect(output).toContain("HP:");
    expect(output).toContain("75/100");
    expect(output).toContain("(75%)");
    expect(output).toContain("█");
    expect(output).toContain("░");
  });

  it("プログレスバーの表示", () => {
    const ProgressBar: React.FC<{ value: number; max: number; label: string }> = ({
      value,
      max,
      label
    }) => {
      const percentage = Math.round((value / max) * 100);
      const width = 20;
      const filled = Math.round((value / max) * width);
      
      return (
        <Box flexDirection="column">
          <Text>{label}: {percentage}%</Text>
          <Box>
            <Text>[</Text>
            <Text color="green">{"=".repeat(filled)}</Text>
            <Text>{" ".repeat(width - filled)}</Text>
            <Text>]</Text>
          </Box>
        </Box>
      );
    };

    const { lastFrame } = render(
      <ProgressBar value={30} max={100} label="Loading" />
    );
    
    expect(lastFrame()).toContain("Loading: 30%");
    expect(lastFrame()).toContain("[");
    expect(lastFrame()).toContain("=");
    expect(lastFrame()).toContain("]");
  });

  it("スナップショット - プログレスバー", () => {
    const ProgressBar: React.FC<{ value: number; max: number; label: string }> = ({
      value,
      max,
      label
    }) => {
      const percentage = Math.round((value / max) * 100);
      const width = 20;
      const filled = Math.round((value / max) * width);
      
      return (
        <Box flexDirection="column">
          <Text>{label}: {percentage}%</Text>
          <Box>
            <Text>[</Text>
            <Text color="green">{"=".repeat(filled)}</Text>
            <Text>{" ".repeat(width - filled)}</Text>
            <Text>]</Text>
          </Box>
        </Box>
      );
    };

    const { lastFrame } = render(
      <ProgressBar value={75} max={100} label="Progress" />
    );
    
    expect(lastFrame()).toMatchSnapshot();
  });
});