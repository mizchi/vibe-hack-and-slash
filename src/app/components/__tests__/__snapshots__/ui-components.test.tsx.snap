// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UI Components > CommonHeader > スナップショット 1`] = `
"╔══════════════════════════════════════════════════════════════════════════════════════════════════╗
║                                                                                                  ║
║ Hack & Slash - Warrior Lv.5                                              [Gold: 100] [Speed: x3] ║
║                                                                                                  ║
║ Tab: [戦闘詳細] | [装備管理]                                                                     ║
║                                                                                                  ║
╚══════════════════════════════════════════════════════════════════════════════════════════════════╝
"
`;

exports[`UI Components > OpeningView > スナップショット 1`] = `
"╔══════════════════════════════════════════════════════════════════════════════════════════════════╗
║                                                                                                  ║
║ ⚔️  Hack & Slash - クラス選択                                                                    ║
║                                                                                                  ║
╚══════════════════════════════════════════════════════════════════════════════════════════════════╝

▶ Warrior: 高い体力と防御力を持つ戦士                       ┌──────────────────────────────────────┐
                                                            │                                      │
  Mage: 強力な魔法と高いマナを持つ魔法使い                  │ Warrior のステータス                 │
                                                            │ HP: 120                              │
  Rogue: 高いクリティカル率と攻撃速度を持つ盗賊             │ 攻撃力: 15                           │
                                                            │ 防御力: 5                            │
  Paladin: 攻撃と防御のバランスが取れた聖騎士               │ マナ: 30                             │
                                                            │                                      │
                                                            └──────────────────────────────────────┘

↑↓: 選択 | Enter: 決定"
`;

exports[`UI Components > Pagination > スナップショット 1`] = `"[1 2 3 4 5 ... 10] (21-30/100)"`;

exports[`UI Helpers > スナップショット - プログレスバー 1`] = `
"Progress: 75%
[===============     ]"
`;
